client/starwhale/api/_impl/dataset/builder.py

```diff
@@ -210,271 +242,202 @@ class SWDSBinBuildExecutor(BaseBuildExecutor):
-            if isinstance(row_data, bytes):
-                _artifact = Binary(row_data, self.default_data_mime_type)
-            elif isinstance(row_data, BaseArtifact):
-                _artifact = row_data
-            else:
-                raise NoSupportError(f"data type {type(row_data)}")
-
-            if not dataset_annotations:
-                # TODO: check annotations type and name
-                dataset_annotations = row_annotations
-
-            _bin_section = self._write(dwriter, _artifact.to_bytes())
-            self.tabular_dataset.put(
-                TabularDatasetRow(
-                    id=idx,
-                    data_link=Link(str(fno)),
-                    data_format=self.data_format_type,
-                    object_store_type=ObjectStoreType.LOCAL,
-                    data_offset=_bin_section.raw_data_offset,
-                    data_size=_bin_section.raw_data_size,
-                    _swds_bin_offset=_bin_section.offset,
-                    _swds_bin_size=_bin_section.size,
-                    data_origin=DataOriginType.NEW,
-                    data_type=_artifact.astype(),
-                    annotations=row_annotations,
-                    _append_seq_id=append_seq_id,
-                )
+            self.dwriter_path = Path(bin_writer_path)
+            self.dwriter = self.dwriter_path.open("wb")
+            self.total_bin_size = 0
+            self._to_update_rows: t.List[TabularDatasetRow] = []
+            self._lock = threading.Lock()
+
+        def __enter__(self) -> SWDSBinBuildExecutor._BinWriter:
+            return self
+
+        def __exit__(
+            self,
+            type: t.Optional[t.Type[BaseException]],
+            value: t.Optional[BaseException],
+            trace: TracebackType,
+        ) -> None:
+            if value:  # pragma: no cover
+                logger.warning(f"type:{type}, exception:{value}, traceback:{trace}")
+
+            self.close()
+
+        def write_row(self, row: TabularDatasetRow) -> None:
+            with self._lock:
+                artifacts = row.artifacts()
+                artifacts_with_bin = False
+                for v in artifacts:
+                    if not v.link and isinstance(v.fp, (str, Path)):
+                        v.link = Link(v.fp, with_local_fs_data=True)
+                    if v.link and v.link.with_local_fs_data:
+                        v.link.uri = self._copy_file(v.link.uri, False)
+                    if (
+                        not v.link
+                        and (isinstance(v.fp, bytes) or isinstance(v.fp, io.IOBase))
+                        and len(v.to_bytes()) > self.bin2link_threshold
+                    ):
+                        artifacts_with_bin = True
+                        _bin_section = self._write(v.to_bytes())
+                        v.link = Link(
+                            self.dwriter_path,
+                            offset=_bin_section.raw_data_offset,
```

Comment:
bin_offset and offset make some confusions in the Link type. bin_size and size have the same issue.

Reply:
> bin_offset and offset make some confusions in the Link type. bin_size and size have the same issue.
`bin_offset` and `bin_size` are key word args for Link.

Reply:
got